File: LogisimPlus/LogisimPlus/logisim/gui/menu/WindowManagers.cs
// ====================================================================================================
// Produced by the Free Edition of Java to C# Converter.
// To produce customized conversions, purchase a Premium Edition license:
// https://www.tangiblesoftwaresolutions.com/product-details/java-to-csharp-converter.html
// ====================================================================================================

using System.Collections.Generic;

/* Copyright (c) 2010, Carl Burch. License information is located in the
 * logisim.Main source code and at www.cburch.com/logisim/. */

namespace logisim.gui.menu
{

	using AnalyzerManager = logisim.analyze.gui.AnalyzerManager;
	using LibraryEvent = logisim.file.LibraryEvent;
	using LibraryListener = logisim.file.LibraryListener;
	using PreferencesFrame = logisim.gui.prefs.PreferencesFrame;
	using Project = logisim.proj.Project;
	using ProjectEvent = logisim.proj.ProjectEvent;
	using ProjectListener = logisim.proj.ProjectListener;
	using Projects = logisim.proj.Projects;
	using WindowMenuItemManager = logisim.util.WindowMenuItemManager;

	public class WindowManagers
	{
		private WindowManagers()
		{
		}

		public static void initialize()
		{
			if (!initialized)
			{
				initialized = true;
				AnalyzerManager.initialize();
				PreferencesFrame.initializeManager();
				Projects.addPropertyChangeListener(Projects.projectListProperty, myListener);
				computeListeners();
			}
		}

		private static bool initialized = false;
		private static MyListener myListener = new MyListener();
		private static Dictionary<Project, ProjectManager> projectMap = new LinkedHashMap<Project, ProjectManager>();

		private class MyListener : PropertyChangeListener
		{
			public virtual void propertyChange(PropertyChangeEvent @event)
			{
				computeListeners();
			}
		}

		private class ProjectManager : WindowMenuItemManager, ProjectListener, LibraryListener
		{
			internal Project proj;

			internal ProjectManager(Project proj) : base(proj.LogisimFile.Name, false)
			{
				this.proj = proj;
				proj.addProjectListener(this);
				proj.addLibraryListener(this);
				frameOpened(proj.Frame);
			}

			public override JFrame getJFrame(bool create)
			{
				return proj.Frame;
			}

			public virtual void projectChanged(ProjectEvent @event)
			{
				if (@event.Action == ProjectEvent.ACTION_SET_FILE)
				{
					Text = proj.LogisimFile.Name;
				}
			}

			public virtual void libraryChanged(LibraryEvent @event)
			{
				if (@event.Action == LibraryEvent.SET_NAME)
				{
					Text = (string) @event.Data;
				}
			}
		}

		private static void computeListeners()
		{
			List<Project> nowOpen = Projects.OpenProjects;

			Dictionary<Project, ProjectManager>.KeyCollection closed = new HashSet<Project>(projectMap.Keys);
			closed.removeAll(nowOpen);
			foreach (Project proj in closed)
			{
				ProjectManager manager = projectMap[proj];
				manager.frameClosed(manager.getJFrame(false));
				projectMap.Remove(proj);
			}

			HashSet<Project> opened = new LinkedHashSet<Project>(nowOpen);
			opened.RemoveAll(projectMap.Keys);
			foreach (Project proj in opened)
			{
				ProjectManager manager = new ProjectManager(proj);
				projectMap[proj] = manager;
			}
		}
	}

}

File: logisim/app/src/main/java/logisim/gui/menu/WindowManagers.java
/* Copyright (c) 2010, Carl Burch. License information is located in the
 * logisim.Main source code and at www.cburch.com/logisim/. */

package logisim.gui.menu;

import java.beans.PropertyChangeEvent;
import java.beans.PropertyChangeListener;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;

import javax.swing.JFrame;

import logisim.analyze.gui.AnalyzerManager;
import logisim.file.LibraryEvent;
import logisim.file.LibraryListener;
import logisim.gui.prefs.PreferencesFrame;
import logisim.proj.Project;
import logisim.proj.ProjectEvent;
import logisim.proj.ProjectListener;
import logisim.proj.Projects;
import logisim.util.WindowMenuItemManager;

public class WindowManagers {
	private WindowManagers() {
	}

	public static void initialize() {
		if (!initialized) {
			initialized = true;
			AnalyzerManager.initialize();
			PreferencesFrame.initializeManager();
			Projects.addPropertyChangeListener(Projects.projectListProperty, myListener);
			computeListeners();
		}
	}

	private static boolean initialized;
	private static MyListener myListener = new MyListener();
	private static HashMap<Project, ProjectManager> projectMap = new LinkedHashMap<>();

	private static class MyListener implements PropertyChangeListener {
		public void propertyChange(PropertyChangeEvent event) {
			computeListeners();
		}
	}

	private static class ProjectManager extends WindowMenuItemManager implements ProjectListener, LibraryListener {
		private Project proj;

		ProjectManager(Project proj) {
			super(proj.getLogisimFile().getName(), false);
			this.proj = proj;
			proj.addProjectListener(this);
			proj.addLibraryListener(this);
			frameOpened(proj.getFrame());
		}

		@Override
		public JFrame getJFrame(boolean create) {
			return proj.getFrame();
		}

		public void projectChanged(ProjectEvent event) {
			if (event.getAction() == ProjectEvent.ACTION_SET_FILE) setText(proj.getLogisimFile().getName());
		}

		public void libraryChanged(LibraryEvent event) {
			if (event.getAction() == LibraryEvent.SET_NAME) setText((String) event.getData());
		}
	}

	private static void computeListeners() {
		List<Project> nowOpen = Projects.getOpenProjects();

		HashSet<Project> closed = new HashSet<>(projectMap.keySet());
		closed.removeAll(nowOpen);
		for (Project proj : closed) {
			ProjectManager manager = projectMap.get(proj);
			manager.frameClosed(manager.getJFrame(false));
			projectMap.remove(proj);
		}

		HashSet<Project> opened = new LinkedHashSet<>(nowOpen);
		opened.removeAll(projectMap.keySet());
		for (Project proj : opened) {
			ProjectManager manager = new ProjectManager(proj);
			projectMap.put(proj, manager);
		}
	}
}
